js/data/buffer.js

```diff
@@ -0,0 +1,289 @@
+        assert(value === null);
+        this.arrayBuffer = null;
+    },
+
+    // binds the buffer to a webgl context
+    bind: function(gl) {
+        var type = gl[this.arrayType];
+        if (!this.buffer) {
+            this.buffer = gl.createBuffer();
+            gl.bindBuffer(type, this.buffer);
+            gl.bufferData(type, this.array.slice(0, this.pos), gl.STATIC_DRAW);
+
+            // dump array buffer once it's bound to gl
+            this.array = null;
+        } else {
+            gl.bindBuffer(type, this.buffer);
+        }
+    },
+
+    destroy: function(gl) {
+        if (this.buffer) {
+            gl.deleteBuffer(this.buffer);
+        }
+    },
+
+    setupViews: function() { /* intentional no-op */ }
+
+};
+// End Buffer1 compatability code
+
+/**
+ * Get an item from the `ArrayBuffer`.
+ * @private
+ * @param {number} index The index of the item to get
+ * @returns {BufferItem}
+ */
+Buffer.prototype.get = function(index) {
+    var item = {};
+    for (var attributeName in this.attributes) {
+        var attribute = this.attributes[attributeName];
+        item[attributeName] = [];
+
+        for (var componentIndex = 0; componentIndex < attribute.components; componentIndex++) {
+            var offset = this.getIndexAttributeOffset(index, attributeName, componentIndex) / attribute.type.size;
+            var arrayBufferView = this.arrayBufferViews[attribute.type.name];
+            var value = arrayBufferView[offset];
+            item[attributeName][componentIndex] = value;
+        }
+    }
+    return item;
+};
+
+/**
+ * Get the byte offset of a particular index.
+ * @private
+ * @param {number} index
+ */
+Buffer.prototype.getIndexOffset = function(index) {
+    return index * this.itemSize;
+};
+
+/**
+ * Get the byte offset of an attribute at a particular item index
+ * @private
+ * @param {number} index
+ * @param {string|BufferAttribute} attribute The attribute to set
+ * @param {number} componentIndex
+ */
+Buffer.prototype.getIndexAttributeOffset = function(index, attributeName, componentIndex) {
+    var attribute = this.attributes[attributeName];
+    return (
+        this.getIndexOffset(index) +
+        attribute.offset +
+        attribute.type.size * (componentIndex || 0)
+    );
+};
+
+Buffer.prototype._resize = function(capacity) {
+    var old = this.arrayBufferViews.UNSIGNED_BYTE;
+    this.capacity = align(capacity, Buffer.CAPACITY_ALIGNMENT);
+    this.arrayBuffer = new ArrayBuffer(this.capacity);
+    this._refreshArrayBufferViews();
+    this.arrayBufferViews.UNSIGNED_BYTE.set(old);
+};
+
+Buffer.prototype._refreshArrayBufferViews = function() {
+    this.arrayBufferViews = {
+        UNSIGNED_BYTE:  new Uint8Array(this.arrayBuffer),
+        BYTE:           new Int8Array(this.arrayBuffer),
+        UNSIGNED_SHORT: new Uint16Array(this.arrayBuffer),
+        SHORT:          new Int16Array(this.arrayBuffer)
+    };
+};
+
+Buffer.prototype._createPushMethod = function() {
```

Comment:
Why do this as its own function and not just part of the Buffer function? Doing it this ways adds both the _createPushMethod and the attributeNames to this?